web-server-lib/web-server/http/request.rkt

```diff
@@ -18,312 +35,402 @@
+      (when (> total-size max-body-length)
+        (network-error 'complete-request "content (~a) exceeds max length (~a)" total-size max-body-length))
+
+      ;; This is safe because of the preceding guard on total-size,
+      ;; though it would be nice if we could somehow avoid the buffering
+      ;; here and write directly from one port to the other.
       (define data-bytes (read-bytes size-in-bytes real-ip))
       (write-bytes data-bytes decode-op)
-      ;; Ignore CRLF
-      (read-line real-ip 'any))
+      (read-http-line/limited real-ip 2))
+
      (define more-headers
-       (list* (header #"Content-Length"
-                      (string->bytes/utf-8 (number->string total-size)))
-              (read-headers real-ip)))
+       (list* (header #"Content-Length" (string->bytes/utf-8 (number->string total-size)))
+              (read-headers real-ip max-remaining-headers max-header-length)))
+
      (close-output-port decode-op)
      (values decoded-ip (append initial-headers more-headers))]
+
     [_
      (values real-ip initial-headers)]))
 
-(define (make-ext:read-request
-         #:connection-close? [connection-close? #f]
-         #:read-timeout [read-timeout 60])
-  (define read-request
-    (make-read-request #:connection-close? connection-close?
-                       #:read-timeout read-timeout))
-  (define (ext:read-request conn host-port port-addresses)
-    (with-handlers ([exn:fail?
-                     (lambda (exn)
-                       (kill-connection! conn)
-                       (raise exn))])
-      (read-request conn host-port port-addresses)))
-  ext:read-request)
-
-(define ext:read-request (make-ext:read-request #:connection-close? #f))
 
 ;; **************************************************
 ;; close-connection?
 
 ; close-connection? : (listof (cons symbol bytes)) number number string string -> boolean
-;; determine if this connection should be closed after serving the
-;; response
+; determine if this connection should be closed after serving the
+; response
 (define close-connection?
   (let ([rx (byte-regexp #"[cC][lL][oO][sS][eE]")])
     (lambda (headers major minor client-ip host-ip)
       (or (< major 1)
           (and (= major 1) (= minor 0))
           (match (headers-assq* #"Connection" headers)
-            [(struct header (f v))
-             (and (regexp-match rx v)
-                  #t)]
-            [#f
-             #f])
+            [(struct header (f v)) (regexp-match? rx v)]
+            [#f #f])
           (msie-from-local-machine? headers client-ip host-ip)))))
 
-;; msie-from-local-machine? : table str str -> bool
-
-;; to work around an error in MSIE for documents < 265 bytes when
-;; connecting from the local machine.  The server could pad the
-;; response as MSIIS does, but closing the connection works, too.  We
-;; do not check for version numbers since IE 6 under windows is 5.2
-;; under macosX
+; msie-from-local-machine? : table str str -> bool
```

Comment:
@jeapostrophe what do you think about dropping this check? This is a very specific problem with a version of IE for which Microsoft themselves dropped support in 2016.

Reply:
Do any other web servers do this?

Reply:
I looked a few servers up:

| Server | Has IE check? |
|-----|------|
| `nginx` | [✅](https://github.com/nginx/nginx/blob/5a2ce3f4ee55eb8903aa9481deaaf402d5a2e805/src/http/ngx_http_core_module.c#L1261) |
| `httpd` (Apache) | 🚫 |
| `httpd` (OpenBSD) | 🚫 |
| `lighttpd` | 🚫 |
| `gunicorn` | 🚫 |
| `webrick` |  🚫 |

I just looked up the source for all of these on github and did a search for `MSIE` so this was in no way exhaustive.

A few more datapoints:

```
$ for url in 'https://google.com?q=racket' 'https://twitter.com' 'https://nytimes.com' 'https://bbc.co.uk' 'https://reddit.com'
    curl -v -H'User-agent: Mozilla/5.0 (compatible; MSIE 6.0; Windows NT 5.1)' "$url" 2>&1 | grep 'left intact'
end
* Connection #0 to host google.com left intact
* Connection #0 to host twitter.com left intact
* Connection #0 to host nytimes.com left intact
* Connection #0 to host bbc.co.uk left intact
* Connection #0 to host reddit.com left intact
```

Although this check is specifically for MSIE connections from localhost so it's not exactly the same thing.

Reply:
I love deleting weird cases like this and I feel pretty safe doing it if Apache does.